tools/schemacode/bidsschematools/render/text.py

```diff
@@ -480,3 +480,37 @@ def define_common_principles(schema, src_path=None):
     string = string.replace("SPEC_ROOT", utils.get_relpath(src_path))
 
     return string
+
+
+def define_allowed_top_directories(schema, src_path=None) -> str:
+    """Create a list of allowed top-level directories with their descriptions.
+
+    Parameters
+    ----------
+    schema : dict
+        The BIDS schema.
+    src_path : str | None
+        The file where this macro is called, which may be explicitly provided
+        by the "page.file.src_path" variable.
+
+    Returns
+    -------
+    string : str
+        Unordered list describing top level directories.
+    """
+
+    string = ""
+
+    for key in schema["objects"]["files"]:
+
+        if schema["objects"]["files"][key]["file_type"] == "directory":
+
+            level = schema["rules"]["files"]["common"]["core"][key]["level"]
+
+            description = schema["objects"]["files"][key]["description"]
+
+            string += f"- `{key}`: [{level.upper()}] {description}"
+
+    string = string.replace("SPEC_ROOT", utils.get_relpath(src_path))
+
+    return string
```

Comment:
1) Level seems redundant (and a bit distracting in the render). Using `MAY` above should do the trick.
2) A lot of redundant lookups in the schema.

Using the schema as a namespace and dropping level, we can do this more tersely:

```suggestion
    string = ""

    for dirname, definition in schema.objects.files.items():
        if definition.file_type == "directory":
            string += f"- `{dirname}`: {definition.description}"

    return string.replace("SPEC_ROOT", utils.get_relpath(src_path))
```

Reply:
ha thanks I had not yet looked up how to use the namespaces you have implemented in the schema.

concerning "level" I do not feel strongly about it, I think I was prematurely optimizing for some weird cases where the level value would not be optional.

Reply:
> I had not yet looked up how to use the namespaces you have implemented in the schema.

The short version is "you can use it as a dict, or an object; if the result is a dict, it will also be a namespace".

Reply:
did you put this as TL;DR in the readme :-p ?